contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,185 @@
+
+* `LocalVolumes` - A collection of volumes mounted in the local storage directory on the node. The discoverer is triggered to acquire this information.
+* `DeleteOperations` - A list keeping track of information related to each delete operation. This status is managed by the cleaner. It contains the following fields:
+  * PV - The PV to be deleted on disk.
+  * DeleteStatus - takes one of the following values:
+    * Requested
+    * InProgress
+    * Complete
+    * Error
+    
+There exists one ProvisionerStatus object per node.
+
+To ensure correctness, master and worker never write to the same field. Specifically:
+* After a `DeleteOperation` is created, its `PV` field is never modified.
+* Master can only create and delete `DeleteOperation`s. It never modifies those objects.
+* Worker can only modify `DeleteStatus`. It can never create or delete `DeleteOperation`s.
+* Only a worker can write to `LocalVolumes`.
+
+### Worker Provisioner
+Create the ProvisionerStatus object on initialization.
+
+The Discoverer periodically probes the filesystem and updates ProvisionerStatus `LocalVolumes`. This can be further improved by using a filesystem watch.
+
+The Deleter uses ProvisionerStatus to keep track of states for each PV being deleted. Deletion is started when the ProvisionerStatus is updated (by the master) and some PVs are in the Requested state. Details will be included in the async volume cleaner design.
+
+### Master Provisioner
+A single goroutine processes events from all informers.
+
+On ProvisionerStatus Informer update, perform the following actions:
+1. For deletion: If a volume is in the Completed state, delete the corresponding PV from the API server and remove the entry from `DeleteOperations`.
+1. For discovery: compare `LocalVolumes` with existing PVs in the API server. If a volume exists in `LocalVolumes` but has not been persisted, and if the disk is not about to be deleted (i.e. in Requested or InProgress state), then create the PV.
+
+On PV Informer update: if the updated PV is in VolumeReleased state, add the PV to `DeleteOperations` with the Requested state.
+
+On Node deletion: delete the associated ProvisionerStatus object.
+
+**TODO**: Consider marking cache dirty when its corresponding API object is modified within master and ignoring Informer events if cache is dirty. This ensures master is always operating on the most up-to-date data. 
+
+## Design 4: Hybrid RPC-CRD Approach
+Keep the same CRD object described in Design 3, but only master has write access to them. Workers must persist its delete state locally and report its status to master through gRPC. Master makes requests to workers by writing to CRD, and fetches worker status by polling (similar to Design 1).
+
+Several modifications to Design 1:
+* For discovery, do not rely on worker to report its identity as it can be compromised. Identify workers using the Endpoints object (pod name and node name).
```

Comment:
Chatted with @thockin - something like this might be possible by providing a link-local kubelet address, which might have other uses as well. However, I've convinced myself that this specific use case (node attestation) is not a good one. (I.e. we should opt for something like mutual TLS or HostPort instead).

Reply:
Ideally the pod identity doc has sufficiently trustworthy provenance and asserts sufficiently informative claims for a recipient to such an authorization decision. If pod proof of identity is not strong enough, we should invest in making it stronger instead of building other forms of authentication.